### PR TITLE
Finalize documentation

### PR DESCRIPTION
**Pull request by nnistelrooij**
Code was already done, but I had one more look at the documentation and finished it.
This should be the final pull request of this project.

### CODE REVIEW

File: nn.py [hunk 1]
```diff
@@ -21,15 +21,15 @@ def __init__(self, device):
         self.device = device
 
     def forward(self, input, target):
-        """Sums cross entropy losses of given predictions and targets.
+        """Sums cross-entropy losses of given predictions and targets.
 
         Args:
             input  = [tuple] sequence of tensors of (raw) predictions
             target = [tuple] sequence of tensors of targets
```
Comment (rusane) [inline nn.py]: For consistency also call it hard targets here like in `LabelSmoothingLoss`:  
target = [tuple] sequence of tensors of **hard** targets

File: nn.py [hunk 2]
```diff
@@ -21,15 +21,15 @@ def __init__(self, device):
...
 
         Returns [torch.Tensor]:
             The grapheme_root, vowel_dacritic, consonant_diacritic,
-            and combined losses given the predictions and targets.
+            and combined losses, given the predictions and targets.
```
Comment (rusane) [inline nn.py]: Same as on line 28 call it hard targets:
... and combined losses, given the predictions and **hard** targets.

File: nn.py [hunk 3]
```diff
@@ -113,31 +112,48 @@ def _split_vectors(vectors, num_augments):
...
     def __init__(self, device, kernel_size=3):
+        """Initialize the simple convolutional neural network.
+
+        Args:
+            device      = [torch.device] device to compute the predictions on
+            kernel_size = [int] kernel size for the two convolutional layers
+        """
         super(ZeroNet, self).__init__()
 
         # images are 128 * 128
```
Comment (rusane) [inline nn.py]: Remove this comment as they can also be 64 * 64.

File: nn.py [hunk 4]
```diff
@@ -178,10 +196,15 @@ def forward(self, x, num_augments=None):
...
-    """Model that uses pre-trained ResNet-50 for intermediate layers.
+    """Model that uses pre-trained ResNet-18 for intermediate layers.
 
     Attributes:
-        device = [torch.device] device to run the model on
+        conv1    = [nn.Module] first convolutional layer
+        resnet18 = [nn.Module] non-pretrained layers of ResNet-18 architecture
+        fc2      = [nn.Module] first fully-connected layer
+        fc3      = [nn.Module] second fully-connected layer
+        fc4      = [nn.Module] third fully-connected layer
```
Comment (rusane) [inline nn.py]: Probably better to call them `fc1`, `fc2`, and `fc3`.

File: nn.py [hunk 5]
```diff
@@ -199,6 +222,7 @@ def __init__(self, device):
         self.fc3 = nn.Linear(512, 11)
         self.fc4 = nn.Linear(512, 7)
```
Comment (rusane) [inline nn.py]: Same as above: call these `fc1`, `fc2`, and `fc3`. (couldn't select the line with `self.fc2` in review mode)

File: nn.py [hunk 6]
```diff
@@ -207,23 +231,23 @@ def forward(self, x, num_augments=None):
...
-        # get latent vectors from ResNet50
+        # get image embeddings from ResNet-18
         h = self.resnet18(h)
         h = h.flatten(start_dim=1)
 
-        # determine sub-problem logits
+        # determine predictions for each sub-problem
         h_graph, h_vowel, h_conso = _split_vectors(h, num_augments)
         y_graph = self.fc2(h_graph)
         y_vowel = self.fc3(h_vowel)
```
Comment (rusane) [inline nn.py]: Same as above: call these `fc1`, `fc2`, and `fc3`. (couldn't select the line with `self.fc4` in review mode)

File: optim.py [hunk 7]
```diff
@@ -143,9 +143,9 @@ def optimize(model,
         val_loader    = [DataLoader] validation data loader
         val_writer    = [SummaryWriter] TensorBoard writer of validation metrics
```
Comment (rusane) [inline optim.py]: Type should be `MetricWriter` for `val_writer` I believe as all the other writers are of this type.

**Review by rusane — COMMENTED** [review]
I could not put a review comment here: in `utils/transforms.py` on line 8 the docstring should probably say **inherit** instead of inherent.

Overall just some minor remarks (see review comments). I did not change anything as I'd like to see if you agree with the suggestions. If you agree with the suggestions, then I'll make the necessary changes if they haven't been already made.

**Review by MauiGonzo — APPROVED** [review]
good comments!

File: nn.py (hunk 6 above)
Comment (nnistelrooij) [reply nn.py]: The problem with these is that the trained model, i.e. *model.pt* is expecting `fc2`, `fc3`, and `fc4` as I changed the documentation only recently. You can manually map the `fc2`, `fc3`, and `fc4` keys in the dictionary when calling `torch.load('model.pt')` to `fc1`, `fc2`, and `fc3` and then saving the dictionary again to *model.pt* to solve this issue.

**Comment by nnistelrooij** [comment]
Do you think providing the best-performing model is sufficient? The slides say that we should give the source code, so I guess we also need to add the Kaggle notebook.

File: nn.py (hunk 6 above)
Comment (rusane) [reply nn.py]: Ah okay, I'm not sure if that's worth the effort. It's not like it is a major issue so we could just leave it likes this as it is still quite clear what the layers are/mean.

**Comment by rusane** [comment]
~~With providing the model, do you mean to just add the `model.pt` file to the repository? If it is just adding the `*.pt` files then we could add them all if you have them but name them accordingly to the conditions described in the report,~~

Do you mean the submission notebook for Kaggle? I guess we could add it, but I don't think it adds important contributions to the code as most of it is already there in separate files.

Edit: I saw that you just commited 1 min before I submitted the comment. In hindsight, adding just the best model is enough as otherwise we would need to change the `fc` layers in all models.

**Comment by nnistelrooij** [comment]
I added the best-performing model with the `fc1`, `fc2`, and `fc3` names and also a small file with the additional code on the Kaggle notebook.

Ready to merge?

**Review by rusane — APPROVED** [review]
👍

**Pull request merged**